python/tvm/testing.py

```diff
@@ -718,31 +844,421 @@ def parametrize_targets(*args):
+    state before a function.  The return value of that fixture
+    function is then accessible by test functions as that accept it as
+    a parameter.
+
+    Fixture functions can accept parameters defined with
+    :py:func:`tvm.testing.parameter`.
+
+    By default, the setup will be performed once for each unit test
+    that uses a fixture, to ensure that unit tests are independent.
+    If the setup is expensive to perform, then the
+    cache_return_value=True argument can be passed to cache the setup.
+    The fixture function will be run only once (or once per parameter,
+    if used with tvm.testing.parameter), and the same return value
+    will be passed to all tests that use it.  If the environment
+    variable TVM_TEST_DISABLE_CACHE is set to a non-zero value, it
+    will disable this feature and no caching will be performed.
+
+    Example
+    -------
+    >>> @tvm.testing.fixture
+    >>> def cheap_setup():
+    >>>     return 5 # Setup code here.
+    >>>
+    >>> def test_feature_x(target, dev, cheap_setup)
+    >>>     assert(cheap_setup == 5) # Run test here
+
+    Or
+
+    >>> size = tvm.testing.parameter(1, 10, 100)
+    >>>
+    >>> @tvm.testing.fixture
+    >>> def cheap_setup(size):
+    >>>     return 5*size # Setup code here, based on size.
+    >>>
+    >>> def test_feature_x(cheap_setup):
+    >>>     assert(cheap_setup in [5, 50, 500])
+
+    Or
+
+    >>> @tvm.testing.fixture(cache_return_value=True)
+    >>> def expensive_setup():
+    >>>     time.sleep(10) # Setup code here
+    >>>     return 5
+    >>>
+    >>> def test_feature_x(target, dev, expensive_setup):
+    >>>     assert(expensive_setup == 5)
+
+    """
+
+    force_disable_cache = bool(int(os.environ.get("TVM_TEST_DISABLE_CACHE", "0")))
+    cache_return_value = cache_return_value and not force_disable_cache
+
+    # Deliberately at function scope, so that caching can track how
+    # many times the fixture has been used.  If used, the cache gets
+    # cleared after the fixture is no longer needed.
+    scope = "function"
+
+    def wraps(func):
+        if cache_return_value:
+            func = _fixture_cache(func)
+        func = pytest.fixture(func, scope=scope)
+        return func
+
+    if func is None:
+        return wraps
+
+    return wraps(func)
+
+
+def _fixture_cache(func):
+    cache = {}
+
+    # Can't use += on a bound method's property.  Therefore, this is a
+    # list rather than a variable so that it can be accessed from the
+    # pytest_collection_modifyitems().
+    num_uses_remaining = [0]
+
+    # Using functools.lru_cache would require the function arguments
+    # to be hashable, which wouldn't allow caching fixtures that
+    # depend on numpy arrays.  For example, a fixture that takes a
+    # numpy array as input, then calculates uses a slow method to
+    # compute a known correct output for that input.  Therefore,
+    # including a fallback for serializable types.
+    def get_cache_key(*args, **kwargs):
```

Comment:
is this guaranteed to be deterministic? `pickle.dumps` and maybe `hash` with stuff like dicts might not be...though maybe the dict thing is fixed now?

in any case, tbh i think this is pretty complicated for a cache key function. since we are trying to use this with parameterizable test cases, can't we just whitelist types that have an obvious, stable conversion to a cache key, and then error on the rest? i am not going to ever run `python tests/python/unittest/test_bar.py --param=<pickled data>`.

Reply:
For `hash`, it is guaranteed to be deterministic, but `pickle.dumps` is not.  For numpy arrays, `pickle.dumps` is, but that isn't guaranteed across all types.

The difficult part here is that the caching should work for fixtures that are based on other fixtures.  For example, consider the following case.  If we want to cache `correct_output`, then the cache needs to be based on the `input_data` argument.  I agree that I don't think anybody will ever input pickled data from the command line, but this pattern of comparing to the correct output feels like it would be pretty common.

``` python
arr_size = tvm.testing.parameter(1, 16, 256)

@tvm.testing.fixture
def input_data(arr_size):
    return np.random.uniform(size=arr_size)

@tvm.testing.fixture
def correct_output(input_data):
    run_very_slow_method(input_data)

def test_func(target, dev, input_data, correct_output):
    output = func(target, dev, input_data)
    tvm.testing.assert_allclose(target, dev, correct_output)
```

The other scheme I considered was to look up which parameters were indirectly involved in computing a particular fixture and caching based on that parameter or parameters.  In this case, `correct_output` is indirectly based on `arr_size`.  However, that would have introduced a potential failure mode if `correct_output` is cached but `input_data` is not.  In that case, the second target to use `arr_size==1` would look up the cached version `correct_output` associated with `arr_size==1`, but would generate a new random value for `input_data`.  This felt like a worse failure mode than the current one of repeating the fixture setup, which is why I used `pickle.dumps` as the fallback.